.github/workflows/tests.yml

```diff
@@ -17,16 +18,36 @@ jobs:
         python-version: ["3.8", "3.9", "3.10"]
 
     steps:
-    - uses: actions/checkout@v2
+    - uses: actions/checkout@v4
     - name: Set up Python ${{ matrix.python-version }}
-      uses: actions/setup-python@v1
+      uses: actions/setup-python@v4
       with:
         python-version: ${{ matrix.python-version }}
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
-        pip install .
+        python -m pip install .[tests]
     - name: Test with pytest
       run: |
-        pip install pytest
-        pytest -v
+        pytest --cov=pyrokinetics --cov-report=xml --cov-report=term tests
+    - name: Upload coverage artifacts
+      uses: actions/upload-artifact@v3
+      with:
+        name: ${{ format('coverage-python-{0}', matrix.python-version) }}
+        path: coverage.xml
+
+  code-coverage:
+    name: Code coverage
+    needs:
+    - pytest
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v4
+    - name: Download coverage report
+      uses: actions/download-artifact@v3
+      with:
+        name: "coverage-python-3.9"
```

Comment:
Is it worth bumping this up to 3.11 to ensure the IMAS test get covered too?

Reply:
Actually it is available for 3.9 so ignore this. We don't even test for 3.11 as is!

I believe that there will be updates to IMAS that will likely only be available for 3.10 but maybe we can cross that bridge we when get there. We could be proactive and switch it now

Reply:
I went with 3.9 as it's the minimum supported version that still uses all the code. If you test with 3.8, code coverage gives you a 0% on the IMAS bits, so it's not a fair representation (there are workarounds to this but they're messy). I know that IMAS doesn't play well with 3.9 though, so I'd be happy to bump it to 3.10. 

I think it'd be a good idea to start testing 3.11 (and maybe even 3.12 if the GitHub runners support it). I was planning to add it in an upcoming rework of some of the packaging/testing/linting/CI tooling, but I suppose it might make sense to do it now while I'm already messing with the testing CI.

Reply:
Happy to include 3.11 and and consider dropping 3.8

Reply:
I'd prefer to hold on to 3.8 for now since it's the built-in Python version for Ubuntu 20.04 and won't reach end-of-life until October next year, but since we're already having to provide a reduced feature set it might make sense to drop 3.8 earlier than that.